back/boxtribute_server/mobile_distribution/crud.py

```diff
@@ -92,6 +95,88 @@ def change_distribution_event_state(distribution_event_id, distribution_event_st
     return distribution_event
 
 
+def set_products_for_packing_list(
+    user_id, distribution_event_id, product_ids_to_add, product_ids_to_remove
+):
+    """
+    Set the products for a packing list.
+    """
+    with db.database.atomic():
+        # Completed Events should not be mutable anymore
+        distribution_event = DistributionEvent.get_by_id(distribution_event_id)
+        if distribution_event.state == DistributionEventState.Completed:
+            raise ModifyCompletedDistributionEvent(
+                desired_operation="set_products_for_packing_list",
+                distribution_event_id=distribution_event.id,
+            )
+
+        # Remove products
+        for product_id in product_ids_to_remove:
+            remove_all_packing_list_entries_from_distribution_event_for_product(
+                user_id, distribution_event_id, product_id
+            )
+
+        # Add new products
+        for product_id in product_ids_to_add:
+            sizes = (
+                Size.select(Size.id)
+                .join(SizeRange)
+                .join(Product)
+                .where(Product.id == product_id)
+            )
+
+            for size in sizes:
+
+                PackingListEntry.get_or_create(
+                    distribution_event=distribution_event_id,
+                    product=product_id,
+                    size=size.id,
+                    defaults={
+                        "number_of_items": 0,
+                        "created_by": user_id,
+                        "created_on": utcnow(),
+                        "state": PackingListEntryState.NotStarted,
+                    },
+                )
+
+        return True
+
+
+def remove_all_packing_list_entries_from_distribution_event_for_product(
+    user_id, distribution_event_id, product_id
+):
+    """
+    Remove all packing list entries from a distribution event for a product.
+    """
+    with db.database.atomic():
+        packing_list_entries = PackingListEntry.select().where(
+            PackingListEntry.distribution_event == distribution_event_id,
+            PackingListEntry.product == product_id,
+        )
+
+        for packing_list_entry in packing_list_entries:
+            packing_list_entry.delete_instance()
```

Comment:
You could instead do a `PackingListEntry.delete().where(...)` in line 152 (don't forget `.execute()` at the end). Probably more efficient to do an SQL bulk delete.